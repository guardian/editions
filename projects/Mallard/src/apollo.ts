/**
 * Setup Apollo framework. At the moment we use Apollo Local State management
 * mechanism and don't fetch data from GraphQL endpoints.
 */
import { ApolloClient } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { SETTINGS_RESOLVERS } from './helpers/settings/resolvers'
import { resolveWeather } from './helpers/location'

/**
 * Resolvers is what Apollo uses to get the value of field that has never been
 * cached before. For example if you have a query `{ foo @client }`, Apollo
 * looks for the field `foo` at the root of the cache. If not available, it'll
 * look for the resolver `Query.foo` (`Query` is just the "namespace" where
 * all queries start).
 *  *
 * Resolvers are functions, which can be asynchonous. They can gather data from
 * local storage (ex. settings), from local APIs (ex. permissions, geolocation),
 * from remote APIs (ex. REST API).
 *
 * Once a resolver returns a value, Apollo will refresh automatically
 * any component that was requiring this in a query via the `useQuery` function.
 */
const RESOLVERS = {
    Query: {
        ...SETTINGS_RESOLVERS,
        weather: resolveWeather,
    },
}

/**
 * The `Link` is the interface Apollo uses to reach GraphQL servers. Since we
 * don't fetch anything from GraphQL endpoints at the moment, we use a no-op
 * `Link` object. We *must* throw to prevent bad queries from carrying through.
 *
 * Apollo would try to use the `Link` if you have a query such as `{ foo }`
 * without the `@client` annotation. In our current use case, `@client` always
 * needs to be specified.
 */
const link = {
    split: () => {
        throw new Error('Not implemented. Did you forget to use `@client`?')
    },
    concat: () => {
        throw new Error('not implemented Did you forget to use `@client`?')
    },
    request: () => {
        throw new Error('not implemented Did you forget to use `@client`?')
    },
}

export const createApolloClient = () =>
    new ApolloClient({
        cache: new InMemoryCache(),
        link,
        resolvers: RESOLVERS,
    })
