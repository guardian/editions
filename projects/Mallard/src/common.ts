export * from '../../Apps/common/src/index';
