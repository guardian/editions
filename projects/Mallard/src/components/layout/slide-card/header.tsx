import { useNavigation } from '@react-navigation/native';
import type { NativeStackNavigationProp } from '@react-navigation/native-stack';
import React from 'react';
import { StyleSheet, TouchableWithoutFeedback, View } from 'react-native';
import type { MainStackParamList } from 'src/navigation/NavigationModels';
import { color } from 'src/theme/color';
import { metrics } from 'src/theme/spacing';
import { Chevron } from '../../chevron';

const styles = StyleSheet.create({
	headerContainer: {
		height: metrics.headerHeight,
		zIndex: 90,
		alignContent: 'stretch',
		justifyContent: 'center',
		backgroundColor: color.background,
		borderColor: color.line,
		borderTopWidth: StyleSheet.hairlineWidth,
		borderBottomWidth: StyleSheet.hairlineWidth,
	},
	headerChevronContainer: {
		flex: 1,
		width: '100%',
		paddingHorizontal: metrics.horizontal,
		justifyContent: 'center',
		alignItems: 'center',
		zIndex: 20,
	},
});

const Header = () => {
	const navigation =
		useNavigation<NativeStackNavigationProp<MainStackParamList>>();
	return (
		<View style={styles.headerContainer}>
			<TouchableWithoutFeedback
				onPress={navigation.goBack}
				accessibilityHint="Go back"
			>
				<View style={styles.headerChevronContainer}>
					<Chevron color={color.text} />
				</View>
			</TouchableWithoutFeedback>
		</View>
	);
};

export { Header };
