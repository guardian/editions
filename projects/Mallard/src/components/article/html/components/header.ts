import { ArticleType, Image as ImageT, Issue } from 'src/common'
import { css, html, px } from 'src/helpers/webview'
import { useImagePath } from 'src/hooks/use-image-paths'
import { Breakpoints } from 'src/theme/breakpoints'
import { color } from 'src/theme/color'
import { metrics } from 'src/theme/spacing'
import { families } from 'src/theme/typography'
import {
    CreditedImage,
    Article,
    MediaAtomElement,
} from '../../../../../../common/src'
import { CssProps, themeColors } from '../helpers/css'
import { breakSides } from '../helpers/layout'
import { Quotes } from './icon/quotes'
import { Line } from './line'
import { Rating } from './rating'
import { SportScore } from './sport-score'
import { renderMediaAtom } from './media-atoms'

export interface ArticleHeaderProps {
    headline: string
    byline?: string
    kicker?: string | null
    image?: CreditedImage | null
    standfirst?: string
    starRating?: Article['starRating']
    sportScore?: Article['sportScore']
    bylineImages?: { cutout?: ImageT }
    bylineHtml?: string
    mainMedia?: MediaAtomElement
}

const outieKicker = (type: ArticleType) => css`
    .header-container[data-type='${type}'] .header-kicker {
        display: inline-block;
        height: 3em;
        margin-top: -3em;
        padding-right: ${metrics.article.sides};
        margin-left: -10em;
        padding-left: 10em;
        border: none;
        z-index: 9;
    }
    .header-container[data-type='${type}'] .header {
        position: relative;
        z-index: 9;
    }
`

const outieHeader = (type: ArticleType) => css`
    .header-container[data-type='${type}'] .header {
        ${breakSides}
        margin-top: -4em;
        padding-top: 0;
        margin-left: -50em;
        padding-left: 50em;
    }
    .header-container[data-type='${type}'] {
        padding-top: 1px;
    }
    @media (max-width: ${px(Breakpoints.tabletVertical)}) {
        .header-container[data-type='${type}'] .header {
            margin-right: 60px;
        }
        .header-container[data-type='${type}'] .header:after {
            margin-right: ${px((60 + metrics.article.sides) * -1)};
        }
    }
    ${outieKicker(type)}
`

export const headerStyles = ({ colors, theme }: CssProps) => css`

    /* prevent clicks on byline links */
    .header a {
        pointer-events: none;
    }

    .header:after {
        background-image: repeating-linear-gradient(
            to bottom,
            ${color.dimLine},
            ${color.dimLine} 1px,
            transparent 1px,
            transparent 4px
        );
        background-repeat: repeat-x;
        background-position: bottom;
        background-size: 1px 16px;
        content: '';
        display: block;
        height: 16px;
        margin: 0;
    }
    @media (min-width: ${px(Breakpoints.tabletVertical)}) {
        .header:after {
            margin-right: ${px(metrics.article.sides * -1)};
        }
    }
    .header {
        padding-top: ${px(metrics.vertical)};
    }
    .header-container-line-wrap,
    .header-container {
        position: relative;
        -webkit-user-select: none;
        -webkit-user-drag: none;
    }
    .header-container-line-wrap {
        z-index: 100;
        ${breakSides}
    }
    .header-bg {
        left: -50em;
        right: -50em;
        top: 0;
        bottom: 0;
        position: absolute;
        z-index: -1;
    }
    .header-image {
        height: 10%;
        width: 100%;
        object-fit: cover;
        display: block;
        z-index: 99;
        position: relative;
    }
    .header-image > .rating, .sport-score {
        position: absolute;
        bottom:0;
        left:0;
    }
    .header-image.header-image--immersive {
        margin: 0 ${px(metrics.article.sides * -1)};
        width: calc(100% + ${px(metrics.article.sides * 2)});
        padding-top: 100%;
    }
    @media (max-width: ${px(Breakpoints.tabletVertical)}) {
        .header-image.header-image--immersive {
            padding-top: 140%;
        }
    }

    .header-kicker {
        font-family: ${families.titlepiece.regular};
        font-size: 16px;
        line-height: 1;
        color: ${colors.main};
        padding: 0.25em 0 0.8em;
        border-bottom: 1px solid ${color.dimLine};
        display: block;
    }
    .header h1 {
        font-size: 30px;
        font-family: ${families.headline.regular};
        font-weight: 400;
        line-height: 1.125em;
        margin: 0.1em 1em 0.75em 0;
        word-wrap: none;
    }

    @media (min-width: ${px(Breakpoints.tabletVertical)}) {
        .header h1 {
            font-size: 40px;
        }
    }

    .header h1 svg {
        height: .7em;
        transform: scale(1.1);
        width: auto;
        fill: ${colors.main};
    }

    .header-standfirst {
        font-weight: 600;
        color: ${colors.main};
    }

    .header-byline {
        font-size: 16px;
        line-height: 1.25em;
        margin: 0;
    }

    .header-byline-italic {
        font-style: italic;
    }

    .header-byline:not(:empty) {
        padding: 0.25em 0 1.5em;
        position: relative;
    }

    .header-byline > span > a {
        font-style: normal !important;
        text-decoration: none;
        font-weight: 600;
        color: ${colors.main};
    }

    .header-top-byline > a {
        font-style: normal !important;
        text-decoration: none;
    }

    .header-byline:not(:empty):after {
        content: '';
        display: block;
        height: 1px;
        background-color: ${color.dimLine};
        position: absolute;
        bottom: -1px;
        left: ${px(metrics.article.sides * -1)};
        right: ${px(metrics.article.sides * -1)};
    }
    @media (min-width: ${px(Breakpoints.tabletVertical)}) {
        .header-byline:not(:empty):after {
            left: 0;
        }
    }

    .header-top p {
        font-family: ${families.headline.medium};
        letter-spacing: 0.2px;
        line-height: 1.1875em;
        margin-bottom: 0.875em;
        font-size: 18px;
    }

    @media (min-width: ${px(Breakpoints.tabletVertical)}) {
        .header-top p {
            font-size: 18px;
        }
    }

    .header-container:after {
        content: '';
        display: block;
        height: 0;
        margin: 0 -50em;
    }
    .header-opinion-flex {
        display: flex;
        align-items: flex-end;
    }

    .header-opinion-flex > :first-child {
        flex: 1 1 0;
    }

    .header-opinion-flex > :last-child {
        width: 15%;
        overflow: visible;
        flex: 0 0 auto;
    }

    .header-opinion-flex > :last-child img {
        width: 240%;
        display: block;
        float: right;
    }

    .image-as-bg {
        display: block;
        padding-top: 60%;
        background-size: cover;
        background-position: center;
        position: relative;
    }

    .image-as-bg[data-preserve-ratio=true] {
        padding-top:0;
        overflow: visible;
        height: auto;
    }
    .image-as-bg__img {
        width: 100%;
        z-index: 0;
        position: relative;
    }

    .image-as-bg-info {
        position: absolute;
        top:0;
        left:0;
        bottom:0;
        right:0;
        padding: ${px(metrics.vertical)} ${px(metrics.horizontal)};
        color: ${color.palette.neutral[100]};
        background: rgba(20,20,20,.8);
        font-family: ${families.sans.regular};
        z-index: 1;
        display:none;
    }

    .image-as-bg[data-open=true] .image-as-bg-info {
        display:block;
    }

    .image-as-bg > button {
        position: absolute;
        bottom: ${px(metrics.vertical)};
        right: ${px(metrics.horizontal)};
        z-index: 2;
        font-family: ${families.icon.regular};
        background-color: ${colors.main};
        color: ${color.textOverDarkBackground};
        border:none;
        width: ${metrics.fronts.sliderRadius * 2}px;
        height: ${metrics.fronts.sliderRadius * 2}px;
        display: block;
        line-height: .9;
        text-align: center;
        font-size: 1.2em;
        border-radius: 100%;
    }

    .share-button {
        float: right;
        width: ${metrics.buttonHeight};
        height: ${metrics.buttonHeight};
        margin: 0 0 0 8px;
        border: 1px solid black;
        border-radius: 1000px;
        padding: 9px 0 0 0;
        text-align: center;
    }

    .clearfix {
        clear: both;
    }

    /*review*/
    .header-container[data-type='review']:after {
        border-bottom: 1px solid ${color.dimLine};
    }
    .header-container[data-type='review'] .header-bg {
        background-color: ${colors.faded};
    }
    .header-container[data-type='review'] h1 {
        color: ${colors.dark};
        font-family: ${families.headline.bold};
    }
    .header-container[data-type='review'] .header-kicker {
        color: ${colors.dark};
    }
    .header-container[data-type='review'] .header-byline a {
        color: ${colors.dark};
    }
    .header-container[data-type='review'] p {
        color: ${colors.dark};
    }

    /*opinion*/
    .header-container[data-type='opinion']:after {
        border-bottom: 1px solid ${color.dimLine};
    }
    .header-container[data-type='opinion'] .header-bg {
        background-color: ${color.palette.opinion.faded};
    }
    .header-container[data-type='opinion'] .header-kicker {
        display: none;
    }
    .header-container[data-type='opinion'] .header-byline {
        color: ${color.palette.neutral[46]};
    }
    .header-container[data-type='opinion'] h1 {
        font-family: ${families.headline.light};
    }
    .header-container[data-type='opinion'] h1 .header-top-byline {
        color: ${colors.main};
        display: block;
        font-family: ${families.titlepiece.regular};
    }


    /*analysis*/
    .header-container[data-type='analysis']:after {
        border-bottom: 1px solid ${color.dimLine};
    }
    .header-container[data-type='analysis'] .header-bg {
        background-color: ${color.palette.neutral[93]};
    }
    .header-container[data-type='analysis'] .header-kicker {
        display: none;
    }
    .header-container[data-type='analysis'] .header-byline {
        color: ${color.palette.neutral[46]};
    }
    .header-container[data-type='analysis'] h1 {
        font-family: ${families.headline.light};
    }
    .header-container[data-type='analysis'] h1 .header-top-headline {
        text-decoration: underline;
        text-decoration-color: ${colors.main};
        text-decoration-thickness: 1px;
    }
    .header-container[data-type='analysis'] h1 .header-top-byline {
        color: ${colors.main};
        display: block;
        font-family: ${families.titlepiece.regular};
    }

    /*gallery*/
    .header-container[data-type='${
        ArticleType.Gallery
    }'] .header-byline  > span > a {
        color: ${themeColors(theme).text};
    }
    .header-container[data-type='${ArticleType.Gallery}'] h1 {
        font-family: ${families.titlepiece.regular};
        min-height: 2em;
    }

    /*immersive*/
    ${outieHeader(ArticleType.Immersive)}
    .header-container[data-type='immersive'] .header-bg {
        background-color: ${color.palette.neutral[100]};
    }
    .header-container[data-type='immersive'] .header {
        background-color: ${color.palette.neutral[100]};
    }
    .header-container[data-type='immersive'] .header-kicker {
        display: none;
    }
    .header-container[data-type='immersive'] .header-top h1 {
        font-family: ${families.titlepiece.regular};
        color: ${colors.dark};
    }
    .header-container[data-type='immersive'] .header-byline a {
        color: ${colors.dark};
    }

    /*longread*/
    ${outieHeader(ArticleType.Longread)}
    .header-container[data-type='longread'] {
        color: ${color.textOverDarkBackground};
    }
    .header-container[data-type='longread'] .header-bg {
        background-color: ${color.palette.neutral[7]};
    }
    .header-container[data-type='longread'] .header {
        background-color: ${color.palette.neutral[7]};
    }
    .header-container[data-type='longread'] .header-kicker {
        background-color: ${colors.main};
        color: ${color.textOverDarkBackground};
        font-family: ${families.headline.bold};
    }
    .header-container[data-type='longread'] .header-top h1 {
        font-family: ${families.titlepiece.regular};
    }
    .header-container[data-type='longread'] .header-byline {
        color: ${color.textOverDarkBackground};
    }


    /*obit*/
    ${outieKicker(ArticleType.Obituary)}
    .header-container[data-type='${ArticleType.Obituary}'] {
        color: ${color.textOverDarkBackground};
    }
    .header-container[data-type='${ArticleType.Obituary}'] .header-bg {
        background-color: ${color.palette.neutral[20]};
    }
    .header-container[data-type='${ArticleType.Obituary}'] .header {
        background-color: ${color.palette.neutral[20]};
    }
    .header-container[data-type='${ArticleType.Obituary}'] .header-kicker {
        background-color: ${color.palette.neutral[20]};
        color: ${color.textOverDarkBackground};
        font-family: ${families.headline.bold};
    }
    .header-container[data-type='${ArticleType.Obituary}'] .header-top h1 {
        font-family: ${families.titlepiece.regular};
    }
    .header-container[data-type='${ArticleType.Obituary}'] .header-byline,
    ${'' /* this is needed to be more specific than an above style */}
    .header-container[data-type='${ArticleType.Obituary}'] .header-byline a {
        color: ${color.textOverDarkBackground};
    }
`

const Image = ({ image, className }: { image: ImageT; className?: string }) => {
    const path = useImagePath(image)
    return html`
        <img class="${className}" src="${path}" />
    `
}

const MainMediaImage = ({
    image,
    className,
    children,
    preserveRatio,
}: {
    image: CreditedImage
    className?: string
    children?: string
    preserveRatio?: boolean
}) => {
    const path = useImagePath(image)

    return html`
        <div
            class="image-as-bg ${className}"
            data-preserve-ratio="${preserveRatio || 'false'}"
            style="background-image: url(${path}); "
            data-open="false"
        >
            ${preserveRatio &&
                html`
                    <img class="image-as-bg__img" src="${path}" aria-hidden />
                `}
            <button
                aria-hidden
                onclick="this.parentNode.dataset.open = !JSON.parse(this.parentNode.dataset.open)"
            >
                
            </button>
            <div class="image-as-bg-info">
                ${image.caption} ${!image.displayCredit ? '' : image.credit}
            </div>
            ${children}
        </div>
    `
}

const isImmersive = (type: ArticleType) =>
    type === ArticleType.Immersive ||
    type === ArticleType.Longread ||
    type === ArticleType.Obituary ||
    type === ArticleType.Gallery

const hasLargeByline = (type: ArticleType) =>
    type === ArticleType.Opinion || type === ArticleType.Analysis

const Header = ({
    publishedId,
    type,
    ...headerProps
}: {
    showMedia: boolean
    publishedId: Issue['publishedId'] | null
    type: ArticleType
} & ArticleHeaderProps) => {
    const immersive = isImmersive(type)
    const largeByline = hasLargeByline(type)
    const cutout =
        type === ArticleType.Opinion &&
        headerProps.bylineImages &&
        headerProps.bylineImages.cutout
    return html`
        ${immersive &&
            headerProps.image &&
            publishedId &&
            MainMediaImage({
                image: headerProps.image,
                className: 'header-image header-image--immersive',
            })}
        <div class="header-container-line-wrap">
            ${Line({ zIndex: 10 })}
            <div class="header-container wrapper" data-type="${type}">
                <header class="header">
                    ${!immersive &&
                        headerProps.image &&
                        publishedId &&
                        MainMediaImage({
                            className: 'header-image',
                            image: headerProps.image,
                            preserveRatio: true,
                            children: headerProps.starRating
                                ? Rating(headerProps)
                                : headerProps.sportScore
                                ? SportScore({
                                      sportScore: headerProps.sportScore,
                                  })
                                : undefined,
                        })}
                    ${headerProps.mainMedia &&
                        (headerProps.showMedia
                            ? renderMediaAtom(headerProps.mainMedia)
                            : null)}
                    ${headerProps.kicker &&
                        html`
                            <span class="header-kicker"
                                >${headerProps.kicker}</span
                            >
                        `}
                    ${largeByline
                        ? html`
                              <section class="header-top">
                                  <div
                                      class="${cutout && `header-opinion-flex`}"
                                  >
                                      <h1>
                                          ${type === ArticleType.Opinion &&
                                              Quotes()}
                                          <span class="header-top-headline"
                                              >${headerProps.headline}</span
                                          >
                                          <span class="header-top-byline"
                                              >${headerProps.bylineHtml}</span
                                          >
                                      </h1>
                                      ${publishedId &&
                                          cutout &&
                                          html`
                                              <div>
                                                  ${Image({
                                                      image: cutout,
                                                  })}
                                              </div>
                                          `}
                                  </div>
                              </section>
                          `
                        : html`
                              <section class="header-top">
                                  <h1>
                                      ${headerProps.headline}
                                  </h1>
                                  <p>${headerProps.standfirst}</p>
                              </section>
                          `}
                </header>

                ${largeByline
                    ? html`
                          <aside class="header-byline header-standfirst">
                              <span>${headerProps.standfirst}</span>
                          </aside>
                      `
                    : html`
                          <aside class="header-byline header-byline-italic">
                              <button
                                  class="share-button"
                                  onclick="window.ReactNativeWebView.postMessage(JSON.stringify({type: 'share'}))"
                              >
                                  ha
                              </button>
                              <span>${headerProps.bylineHtml}</span>
                              <div class="clearfix"></div>
                          </aside>
                      `}
                <div class="header-bg"></div>
            </div>
        </div>
    `
}

export { Header }
