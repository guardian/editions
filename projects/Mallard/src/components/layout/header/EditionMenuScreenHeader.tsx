import React from 'react'
import { Header } from './header'
import { EditionsMenuButton } from 'src/components/EditionsMenu/EditionsMenuButton/EditionsMenuButton'
import { IssueTitle } from 'src/components/issue/issue-title'

export const EditionsMenuScreenHeader = ({
    leftActionPress,
}: {
    leftActionPress: () => void
}) => (
    <Header
        leftAction={<EditionsMenuButton selected onPress={leftActionPress} />}
        layout={'center'}
    >
        <IssueTitle title={`Editions`} />
    </Header>
)
