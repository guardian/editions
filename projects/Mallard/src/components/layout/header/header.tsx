import type { ReactNode } from 'react';
import React from 'react';
import { StatusBar, StyleSheet, View } from 'react-native';
import type { SpecialEditionHeaderStyles } from '../../../../../Apps/common/src';
import { Highlight } from '../../../components/highlight';
import { GridRowSplit } from '../../../components/issue/issue-title';
import { useInsets, useMediaQuery } from '../../../hooks/use-screen';
import { WithAppAppearance } from '../../../theme/appearance';
import { Breakpoints } from '../../../theme/breakpoints';
import { color } from '../../../theme/color';
import { metrics } from '../../../theme/spacing';
import { getFont } from '../../../theme/typography';

const styles = StyleSheet.create({
	background: {
		backgroundColor: color.primary,
		paddingVertical: metrics.vertical,
		justifyContent: 'flex-end',
		flexDirection: 'row',
	},
	backgroundWhite: {
		backgroundColor: color.background,
		borderBottomColor: color.line,
		borderBottomWidth: StyleSheet.hairlineWidth,
		paddingVertical: metrics.vertical,
		justifyContent: 'flex-end',
		flexDirection: 'row',
	},
	height: {
		height: metrics.vertical + getFont('titlepiece', 1.25).lineHeight * 1.5,
	},
	flex: {
		flexDirection: 'row',
	},
	headerSplit: {
		flexDirection: 'row',
		flex: 1,
	},
	headerTitle: {
		marginRight: metrics.horizontal,
	},

	centerWrapper: {
		width: '100%',
		flexDirection: 'row',
		justifyContent: 'space-between',
		alignItems: 'center',
	},
	centerTitle: {
		...StyleSheet.absoluteFillObject,
		alignItems: 'center',
		justifyContent: 'center',
		zIndex: 1,
	},
	leftAction: {
		width: 90,
		zIndex: 2,
		paddingLeft: metrics.horizontal,
	},
	centerAction: {
		zIndex: 2,
		paddingRight: metrics.horizontal,
	},
});

type TouchableHeaderProps =
	| { onPress: () => void; accessibilityHint: string }
	| {};

type HeaderProps = {
	theme?: 'light' | 'default';
	headerStyles?: SpecialEditionHeaderStyles;
	action?: ReactNode;
	leftAction?: ReactNode;
	layout?: 'issue' | 'center';
	alignment?: 'drawer' | null;
	children: ReactNode;
} & TouchableHeaderProps;

const Header = ({
	action,
	theme,
	headerStyles,
	leftAction,
	layout = 'issue',
	alignment = null,
	children,
	...otherProps
}: HeaderProps) => {
	const { top: marginTop } = useInsets();
	const bg = theme === 'light' ? styles.backgroundWhite : styles.background;
	const isTablet = useMediaQuery(
		(width) => width >= Breakpoints.TabletVertical,
	);
	const headerSplitFlex =
		isTablet && alignment !== 'drawer' ? 'flex-end' : 'space-between';

	return (
		<WithAppAppearance value={theme === 'light' ? 'default' : 'primary'}>
			{theme === 'light' && (
				<StatusBar barStyle="dark-content" backgroundColor="#fff" />
			)}
			<View
				style={[
					bg,
					headerStyles && {
						backgroundColor: headerStyles.backgroundColor,
					},
				]}
			>
				{layout === 'issue' ? (
					<GridRowSplit
						proxy={
							<View style={styles.leftAction}>{leftAction}</View>
						}
						style={[{ marginTop }, styles.height]}
					>
						<View
							style={[
								styles.headerSplit,
								{ justifyContent: headerSplitFlex },
							]}
						>
							<View style={styles.headerTitle}>
								{'onPress' in otherProps ? (
									<Highlight
										onPress={otherProps.onPress}
										accessibilityHint={
											otherProps.accessibilityHint
										}
										hitSlop={{
											top: 10,
											bottom: 10,
											left: 10,
											right: 10,
										}}
									>
										{children}
									</Highlight>
								) : (
									children
								)}
							</View>
							<View style={{ paddingRight: metrics.horizontal }}>
								{action}
							</View>
						</View>
					</GridRowSplit>
				) : (
					<View style={{ marginTop, width: '100%' }}>
						<View style={[styles.height, styles.centerWrapper]}>
							<View style={styles.leftAction}>{leftAction}</View>
							<View style={styles.centerAction}>{action}</View>
							<View style={styles.centerTitle}>{children}</View>
						</View>
					</View>
				)}
			</View>
		</WithAppAppearance>
	);
};

export { Header };
