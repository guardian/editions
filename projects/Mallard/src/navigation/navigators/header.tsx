import {
    NavigationContainer,
    NavigationInjectedProps,
    createStackNavigator,
    NavigationParams,
} from 'react-navigation'

import React from 'react'
import { Header } from 'src/components/layout/header/header'
import { Button } from 'src/components/Button/Button'
import { IssueTitle } from 'src/components/issue/issue-title'
import { addStaticRouter } from '../helpers/base'

interface NavigationOptions {
    title?: string
    showHeaderRight?: boolean
    showHeaderLeft?: boolean
}

type NavOrFn =
    | (({ navigation }: { navigation: NavigationParams }) => NavigationOptions)
    | NavigationOptions

const getNavigationOptions = (
    navigation: NavigationParams,
    options: NavOrFn,
): NavigationOptions => {
    if (!options) return {}
    if (typeof options === 'function') {
        return options({ navigation })
    }
    return options
}
const addStaticRouterWithHeader = (
    Navigator: NavigationContainer,
): NavigationContainer => {
    const { router } = Navigator

    const WithHeader = ({ navigation }: NavigationInjectedProps) => {
        const component = router.getComponentForState(navigation.state)
        const options = getNavigationOptions(
            navigation,
            component.navigationOptions,
        )

        const showHeaderLeft =
            options.showHeaderLeft !== undefined ? options.showHeaderLeft : true
        const showHeaderRight =
            options.showHeaderRight !== undefined
                ? options.showHeaderRight
                : false

        return (
            <>
                <Header
                    leftAction={
                        showHeaderLeft ? (
                            <Button
                                icon={'\uE00A'}
                                alt="Back"
                                onPress={() => navigation.goBack(null)}
                            ></Button>
                        ) : null
                    }
                    action={
                        showHeaderRight ? (
                            <Button
                                icon={'\uE04F'}
                                alt="Back"
                                onPress={() => navigation.goBack(null)}
                            ></Button>
                        ) : null
                    }
                    layout={'center'}
                >
                    <IssueTitle
                        title={options.title || navigation.state.routeName}
                    />
                </Header>
                <Navigator navigation={navigation} />
            </>
        )
    }

    return addStaticRouter(Navigator, WithHeader)
}

const createHeaderStackNavigator = (
    routes: Parameters<typeof createStackNavigator>[0],
    options?: Parameters<typeof createStackNavigator>[1],
) =>
    addStaticRouterWithHeader(
        createStackNavigator(routes, { ...options, headerMode: 'none' }),
    )

export { createHeaderStackNavigator, addStaticRouterWithHeader }
