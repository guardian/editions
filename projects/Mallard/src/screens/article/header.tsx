import { useNavigation } from '@react-navigation/native';
import React from 'react';
import { Button, ButtonAppearance } from 'src/components/Button/Button';
import { Header } from 'src/components/layout/header/header';

export const BasicArticleHeader = () => {
	const navigation = useNavigation();
	return (
		<Header
			theme="light"
			leftAction={
				<Button
					appearance={ButtonAppearance.skeleton}
					icon={'\uE00A'}
					alt="Back"
					onPress={() => navigation.goBack(null)}
				></Button>
			}
			layout="center"
		>
			{null}
		</Header>
	);
};
