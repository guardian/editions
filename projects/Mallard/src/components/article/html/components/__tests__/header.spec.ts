import { getStandFirst } from '../header'
import { HeaderType, ArticleType } from 'src/common'

jest.mock('src/components/article/html/components/icon/quotes', () => ({
    Quotes: () => '<Quotes />',
}))

describe('article html Header', () => {
    describe('getStandFirst', () => {
        it('should display LargeByline Header Type with Opinion Article Type', () => {
            const html = getStandFirst(
                HeaderType.LargeByline,
                ArticleType.Opinion,
                { headline: 'Test Headline', bylineHtml: '<p>Test Byline</p>' },
                null,
                () => undefined,
                'news',
            )
            expect(html).toMatchSnapshot()
        })
        it('should display LargeByline Header Type with Opinion Article Type with a cutout space', () => {
            const html = getStandFirst(
                HeaderType.LargeByline,
                ArticleType.Opinion,
                {
                    headline: 'Test Headline',
                    bylineHtml: '<p>Test Byline</p>',
                    bylineImages: {
                        cutout: { source: 'media', path: 'path/to/image' },
                    },
                },
                null,
                () => undefined,
                'news',
            )
            expect(html).toMatchSnapshot()
        })
        it('should display LargeByline Header Type with Opinion Article Type with a cutout image', () => {
            const html = getStandFirst(
                HeaderType.LargeByline,
                ArticleType.Opinion,
                {
                    headline: 'Test Headline',
                    bylineHtml: '<p>Test Byline</p>',
                    bylineImages: {
                        cutout: { source: 'media', path: 'path/to/image' },
                    },
                },
                '1234567880',
                () => undefined,
                'news',
            )
            expect(html).toMatchSnapshot()
        })
        it('should display LargeByline Header Type with alternative Article Type with a cutout image', () => {
            const html = getStandFirst(
                HeaderType.LargeByline,
                ArticleType.Article,
                {
                    headline: 'Test Headline',
                    bylineHtml: '<p>Test Byline</p>',
                    bylineImages: {
                        cutout: { source: 'media', path: 'path/to/image' },
                    },
                },
                null,
                () => undefined,
                'news',
            )
            expect(html).toMatchSnapshot()
        })
        it('should display RegularByline Header Type, show a standfirst with the headline', () => {
            const html = getStandFirst(
                HeaderType.RegularByline,
                ArticleType.Article,
                {
                    headline: 'Test Headline',
                    bylineHtml: '<p>Test Byline</p>',
                    standfirst: 'Test Standfirst',
                },
                null,
                () => undefined,
                'news',
            )
            expect(html).toMatchSnapshot()
        })
        it('should display NoByline Header Type, show only the headline', () => {
            const html = getStandFirst(
                HeaderType.NoByline,
                ArticleType.Article,
                {
                    headline: 'Test Headline',
                    bylineHtml: '<p>Test Byline</p>',
                },
                null,
                () => undefined,
                'news',
            )
            expect(html).toMatchSnapshot()
        })
    })
})
