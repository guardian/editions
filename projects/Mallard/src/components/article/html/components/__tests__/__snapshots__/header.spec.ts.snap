// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`article html Header getStandFirst should display LargeByline Header Type with Opinion Article Type 1`] = `
"
            <section class=\\"header-top\\">
                <div class=\\"\\">
                    
            <h1>
                <span class=\\"header-top-headline\\"
                    ><Quotes />Test Headline
                </span>
                
                        <span class=\\"header-top-byline\\"
                            ><p>Test Byline</p>
                        </span>
                    
            </h1>
        
                    
                </div>
            </section>
        "
`;

exports[`article html Header getStandFirst should display LargeByline Header Type with Opinion Article Type with a cutout image 1`] = `
"
            <section class=\\"header-top\\">
                <div class=\\"header-opinion-flex\\">
                    
            <h1>
                <span class=\\"header-top-headline\\"
                    ><Quotes />Test Headline
                </span>
                
                        <span class=\\"header-top-byline\\"
                            ><p>Test Byline</p>
                        </span>
                    
            </h1>
        
                    
                            <div>
                                
        <img class=\\"\\" src=\\"\\" />
    
                            </div>
                        
                </div>
            </section>
        "
`;

exports[`article html Header getStandFirst should display LargeByline Header Type with Opinion Article Type with a cutout space 1`] = `
"
            <section class=\\"header-top\\">
                <div class=\\"header-opinion-flex\\">
                    
            <h1>
                <span class=\\"header-top-headline\\"
                    ><Quotes />Test Headline
                </span>
                
                        <span class=\\"header-top-byline\\"
                            ><p>Test Byline</p>
                        </span>
                    
            </h1>
        
                    
                </div>
            </section>
        "
`;

exports[`article html Header getStandFirst should display LargeByline Header Type with alternative Article Type with a cutout image 1`] = `
"
            <section class=\\"header-top\\">
                <div class=\\"\\">
                    
            <h1>
                <span class=\\"header-top-headline\\"
                    >Test Headline
                </span>
                
                        <span class=\\"header-top-byline\\"
                            ><p>Test Byline</p>
                        </span>
                    
            </h1>
        
                    
                </div>
            </section>
        "
`;

exports[`article html Header getStandFirst should display NoByline Header Type, show only the headline 1`] = `
"
            <section class=\\"header-top\\">
                
            <h1>
                Test Headline
            </h1>
        
                
            </section>
        "
`;

exports[`article html Header getStandFirst should display RegularByline Header Type, show a standfirst with the headline 1`] = `
"
            <section class=\\"header-top\\">
                
            <h1>
                Test Headline
            </h1>
        
                
        <p
            
        >
            Test Standfirst
        </p>
    
            </section>
        "
`;
